### DISCUSSION TITLE
Suppressible and non-suppressible certificate validation checks

### DISCUSSION THREAD

**Discussion by oliveirars** [eclipse-milo/milo#1341] [General]
Hi guys. 

I've recently had an issue using certificates to authenticate a software (using Eclipse Milo) on an OPC-UA server. 

While I was investigating, I noticed two issues related to certificate validations. According to OPC-UA version 1.03 specification:

- The [APPLICATION_URI](https://github.com/eclipse/milo/blob/13a6a8e7e303fe3c313f8b9e0ffbcfe789278e15/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/util/validation/ValidationCheck.java#L65) check may not be suppressed. Although this validation is included in the list of [NO_OPTIONAL_CHECKS](https://github.com/eclipse/milo/blob/13a6a8e7e303fe3c313f8b9e0ffbcfe789278e15/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/util/validation/ValidationCheck.java#L72C55-L72C73), a client software component could [instantiate](https://github.com/eclipse/milo/blob/13a6a8e7e303fe3c313f8b9e0ffbcfe789278e15/opc-ua-stack/stack-client/src/main/java/org/eclipse/milo/opcua/stack/client/security/DefaultClientCertificateValidator.java#L50) the `DefaultClientCertificateValidator` passing a collection without this check, making this check optional.
 
- The [REVOCATION](https://github.com/eclipse/milo/blob/13a6a8e7e303fe3c313f8b9e0ffbcfe789278e15/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/util/validation/ValidationCheck.java#L49) check may not be suppressed, but it seems not to be handled as so - it is not in the non-optional checks collection, for example. In this case, an external component could pass a set of checks without this one. 

My understanding can be wrong, but it seems the library should handle these as mandatory. Is there some reason to adopt a different strategy?

![image](https://github.com/user-attachments/assets/f652e98f-dcb2-4770-b58d-d388f71f3446)

**Comment by kevinherron** [comment]
In practice I've found it best to leave an escape hatch for applications/implementations to choose to deviate from the spec on purpose to get a connection established.

I'm not sure additional effort to make it impossible to configure invalid validation checks is worth it, though certainly some sane defaults and documentation about what is or isn't mandatory according to the spec should exist.

**Reply by oliveirars** [reply]
Thank you so much for the explanation @kevinherron .

I think this is a nice decision since you let possibilities to configure or even create custom validators. I reported it to hear the aimed purpose, and it is fair enough.
